R/MultivariateMoments.R

```diff
@@ -17,185 +17,1522 @@
+#' Studies, 23(4), 1467-1502.
+#' 
+#' Simaan, Yusif. 1993. Portfolio selection and asset pricing: three-parameter framework. 
+#' Management Science, 39(5), 68-577.
+###keywords ts multivariate distribution models
+#' @examples
+#' 
+#' data(edhec)
+#' 
+#' # structured estimation with constant correlation model
+#' # 'as.mat = F' would speed up calculations in higher dimensions
+#' sigma <- M2.struct(edhec, "CC")
+#' m3 <- M3.struct(edhec, "CC")
+#' m4 <- M4.struct(edhec, "CC")
+#' 
+#' # compute equal-weighted portfolio modified ES
+#' mu <- colMeans(edhec)
+#' p <- length(mu)
+#' ES(p = 0.95, portfolio_method = "component", weights = rep(1 / p, p), mu = mu, 
+#'     sigma = sigma, m3 = m3, m4 = m4)
+#' 
+#' # compare to sample method
+#' sigma <- cov(edhec)
+#' m3 <- M3.MM(edhec)
+#' m4 <- M4.MM(edhec)
+#' ES(p = 0.95, portfolio_method = "component", weights = rep(1 / p, p), mu = mu, 
+#'     sigma = sigma, m3 = m3, m4 = m4)
+#' 
+#' @export M2.struct
+M2.struct <- function(R, struct = c("Indep", "IndepId", "observedfactor", "CC"), f = NULL) {
+  # @author Dries Cornilly
+  #
+  # DESCRIPTION:
+  # computes different strutured covariance estimators
+  #
+  # Inputs:
+  # R         : numeric matrix of dimensions NN x PP
+  # struct    : select the structured estimator
+  #           : Indep   : independent, unequal marginals (Ledoit and Wolf (2004))
+  #           : IndepId : independent, equal marginals (Ledoit and Wolf (2003))
+  #           : observedfactor : 1-factor or multi-factor linear model
+  #           : CC : constant-correlation model of (Ledoit and Wolf (2004))
+  # f         : numeric vector or matrix with factor observations
+  #
+  # Outputs:
+  # covariance matrix
+  
+  X <- coredata(R)
+  
+  # input checking
+  struct <- struct[1]
+  if (NCOL(X) < 2) stop("R must have at least 2 variables")
+  if ((struct == "observedfactor") & is.null(f)) stop("Provide the factor observations")
+  
+  # compute useful variables
+  NN <- dim(X)[1]                                                   # number of observations
+  PP <- dim(X)[2]                                                   # number of assets
+  
+  Xc <- X - matrix(colMeans(X), nrow = NN, ncol = PP, byrow = TRUE) # center the observations
+  margvars <- colMeans(Xc^2)
+  
+  # compute the coskewness matrix
+  if (struct == "Indep") {
+    # independent marginals
+    T2 <- diag(margvars)
+    return (T2)
+    
+  } else if (struct == "IndepId") {
+    # independent and equally distributed marginals
+    T2 <- mean(margvars) * diag(PP)
+    return (T2)
+    
+  } else if (struct == "observedfactor") {
+    # linear factor model
+    if (NCOL(f) == 1) {
+      beta <- apply(Xc, 2, function(a) cov(a, f) / var(f))
+      fc <- f - mean(f)
+      fvar <- mean(fc^2)
+      T2 <- fvar * beta %*% t(beta)
+      diag(T2) <- margvars
+    } else {
+      mod <- stats::lm(X ~ f)
+      beta <- t(mod$coefficients[-1,])
+      fcov <- cov(f) * (n - 1) / n
```

Comment:
`n` is not defined in this scope.  Is it supposed to be an argument to `M2.struct`?  Is it a function of other inputs?